tools/snapshot/snapshot.py

```diff
@@ -0,0 +1,468 @@
+    """
+    log.debug(f"cleaning up rclone snapshot credentials on {machine['ip']} (s{machine['shard']})")
+    success_msg = "RCLONE_CLEANUP_SUCCESS"
+    cmd = f"[ -f {rsync['config_path_on_client']} ] && rm {rsync['config_path_on_client']} && echo {success_msg} " \
+          f"|| [ ! -f {rsync['config_path_on_client']} ] && echo {success_msg} "
+    cleanup_response = _ssh_cmd(machine['user'], machine['ip'], cmd)
+    if success_msg not in cleanup_response:
+        log.error("failed to clean-up rclone config")
+        log.error(f"check response: {cleanup_response.strip()}\n Expected: {success_msg}")
+        raise RuntimeError("failed to clean-up rclone config")
+    log.debug(f"successfully cleaned up rclone snapshot credentials on {machine['ip']} (s{machine['shard']})")
+
+
+def cleanup_rclone_config():
+    """
+    Clean/remove rclone credentials that were added (as dictated by the config).
+
+    Note that this will remove the file at whatever the `config_path_on_client`
+    specifies in the config.
+    """
+    threads, pool = [], ThreadPool()
+    for machine in machines:
+        threads.append(pool.apply_async(_cleanup_rclone_config, (machine,)))
+    for t in threads:
+        t.get()
+
+
+def _derive_db_paths(machine):
+    """
+    Internal function to derive the true harmony DB and
+    rsync harmony DB path on the machine.
+    """
+    db_path_on_machine = f"{machine['db_directory']}/harmony_db_{machine['shard']}"
+    db_rsync_path_on_machine = f"{db_path_on_machine}_rsync"
+    return db_path_on_machine, db_rsync_path_on_machine
+
+
+def _bucket_sync(machine):
+    """
+    Internal function to start bucket sync.
+    Function call will block until bucket sync is done on machine.
+
+    Note the convention used when syncing to bucket.
+    """
+    log.debug(f'starting bucket sync on {machine["ip"]} (s{machine["shard"]})')
+    _, rsync_db_path = _derive_db_paths(machine)
+    db_type = 'full' if condition['is_archival'] else 'pruned'
+    bucket, shard = rsync['snapshot_bin'], machine['shard']
+    unix_time, config = int(time.time()), rsync['config_path_on_client']
+    cmd = f"rclone sync {rsync_db_path} {bucket}/{db_type}/{shard}/harmony_db_{shard}.{unix_time} --config {config}"
+    try:
+        _ssh_cmd(machine['user'], machine['ip'], cmd)
+    except subprocess.CalledProcessError as e:
+        log.error("failed to bucket sync db")
+        raise RuntimeError("failed to bucket sync db") from e
+    log.debug(f'successful bucket sync on {machine["ip"]} (s{machine["shard"]})')
+
+
+def _local_sync(machine):
+    """
+    Internal function to trigger a local sink.
+    Function call will block until local sync is done on machine.
+    """
+    log.debug(f'starting local sync on {machine["ip"]} (s{machine["shard"]})')
+    db_path_on_machine, db_rsync_path_on_machine = _derive_db_paths(machine)
+    cmd = f"rclone sync {db_path_on_machine} {db_rsync_path_on_machine} --transfers 64"
```

Comment:
did you check the return value of rclone? to make sure it is successful? for example, you may hit disk full error.

Reply:
all ssh cmds will raise an exception if remote command exited with a non-zero exit code.